zrml/prediction-markets/src/migrations.rs

```diff
@@ -157,7 +172,148 @@ impl<T: Config> OnRuntimeUpgrade for MigrateMarketPoolsBeforeOpen<T> {
+            return total_weight;
+        }
+
+        total_weight = total_weight.saturating_add(T::DbWeight::get().reads(2));
+        log::info!("Starting storage cleanup of CleanUpStorageForResolvedOrClosedMarkets");
+
+        let dispute_period = T::DisputePeriod::get();
+        let current_block: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
+        let last_dp_end_block = current_block.saturating_sub(dispute_period);
+        type DisputeBlockToMarketIdsTuple<T> =
+            (<T as frame_system::Config>::BlockNumber, BoundedVec<MarketIdOf<T>, CacheSize>);
+        type IterType<T> = PrefixIterator<DisputeBlockToMarketIdsTuple<T>>;
+
+        let market_ids_per_dispute_iterator: IterType<T> =
+            frame_support::migration::storage_key_iter::<_, _, Twox64Concat>(
+                b"PredictionMarkets",
+                b"MarketIdsPerDisputeBlock",
+            );
+
+        let market_ids_tobe_removed_per_dispute: Vec<DisputeBlockToMarketIdsTuple<T>> =
+            market_ids_per_dispute_iterator
+                .filter(|(dispute_start_block, _market_ids)| {
+                    *dispute_start_block <= last_dp_end_block
+                })
+                .collect();
+        for (dispute_start_block, _market_ids) in market_ids_tobe_removed_per_dispute {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+            MarketIdsPerDisputeBlock::<T>::remove(dispute_start_block);
+        }
+
+        let market_ids_per_report_iterator: IterType<T> =
+            frame_support::migration::storage_key_iter::<_, _, Twox64Concat>(
+                b"PredictionMarkets",
+                b"MarketIdsPerReportBlock",
+            );
+
+        let market_ids_tobe_removed_per_report: Vec<_> = market_ids_per_report_iterator
+            .filter(|(dispute_start_block, _market_ids)| *dispute_start_block <= last_dp_end_block)
+            .collect();
+        for (dispute_start_block, _market_ids) in market_ids_tobe_removed_per_report {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+            MarketIdsPerReportBlock::<T>::remove(dispute_start_block);
+        }
+        StorageVersion::new(
+            PREDICTION_MARKETS_NEXT_STORAGE_VERSION_FOR_CLEANUP_STORAGE_FOR_RESOLVED_MARKETS,
+        )
+        .put::<Pallet<T>>();
+        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+
+        log::info!("Completed storage cleanup of CleanUpStorageForResolvedOrClosedMarkets");
+        total_weight
+    }
 
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        Ok(())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> {
+        let dispute_period = T::DisputePeriod::get();
+        let current_block: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
+        let last_dp_end_block = current_block.saturating_sub(dispute_period);
```

Comment:
As far as I can tell, the runtime upgrades trigger _before_ the block's `on_initialize` is called:  https://github.com/paritytech/substrate/blob/8e600ee0515122667cd091d632104708d343ca34/frame/executive/src/lib.rs#L292-L319

This means that in terms of disputing, the chain state lags one block behind during upgrades.

```suggestion
        let last_dp_end_block = current_block.saturating_sub(dispute_period).saturating_sub(1);
```